### PR TITLE
Add fleet checkpoint on local fs and remote fs(such as hdfs) for EDL

### PR DESCRIPTION
**Pull request by gongweibao**
This PR add checkpoint function for EDL:
1. Save/load checkpoint to/from locals/remote.
2. Add some process control utils.

### CODE REVIEW

**Review by guru4elephant — COMMENTED** [review]
Could you submit the python code into edl repo? and make EDL as a plugin package to use.

**Review by guru4elephant — CHANGES_REQUESTED** [review]
Could you submit the python code into edl repo? and make EDL as a plugin package to use.

**Comment by gongweibao** [comment]
OK! I will split them into the suitable location.

File: paddle/fluid/framework/io/shell.cc [hunk 1]
```diff
@@ -296,23 +299,67 @@ std::pair<std::shared_ptr<FILE>, std::shared_ptr<FILE>> shell_p2open(
 #endif
 }
 
-std::string shell_get_command_output(const std::string& cmd) {
+class ElapsedTime {
```
Comment (gavin1332) [inline paddle/fluid/framework/io/shell.cc]: refer to platform::Timer

File: paddle/fluid/framework/io/shell.cc [hunk 2]
```diff
@@ -296,23 +299,67 @@ std::pair<std::shared_ptr<FILE>, std::shared_ptr<FILE>> shell_p2open(
...
-  return "";
+  PADDLE_THROW(platform::errors::Unimplemented(
+      "This function(shell_get_command_output) is not implemented under _WIN32 "
+      "or __APPLE__."));
 #else
   int err_no = 0;
+  ElapsedTime elapsed;
   do {
+    if (print_cmd) {
+      printf("exec cmd back:%s\n", cmd.c_str());
```
Comment (gavin1332) [inline paddle/fluid/framework/io/shell.cc]: why not use VLOG

File: python/paddle/distributed/cloud_utils.py [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+                logger.warning("Use Cloud specified port:{}.".format(
+                    paddle_port))
+                started_port = paddle_port
+
+        except Exception as e:
+            print(e)
+            pass
+
+    if started_port is None:
+        started_port = 6170
```
Comment (gavin1332) [inline python/paddle/distributed/cloud_utils.py:67]: why not detecting an available port?

File: python/paddle/distributed/fs_wrapper.py [hunk 4]
```diff
@@ -0,0 +1,223 @@
...
+    def mkdir(self, fs_path):
+        assert not os.path.isfile(fs_path), "{} is already a file".format(
+            fs_path)
+        os.system("mkdir -p {}".format(fs_path))
+
+    def mv(self, fs_src_path, fs_dst_path):
+        os.rename(fs_src_path, fs_dst_path)
+
+    def rmr(self, fs_path):
+        shutil.rmtree(fs_path)
```
Comment (gavin1332) [inline python/paddle/distributed/fs_wrapper.py]: incompatible to python2.7

File: python/paddle/distributed/fs_wrapper.py [hunk 5]
```diff
@@ -0,0 +1,223 @@
...
+
+import paddle.fluid as fluid
+import sys
+import abc
+import os
+from pathlib import PurePosixPath
+import shutil
+
+
+class FS(object):
```
Comment (gavin1332) [inline python/paddle/distributed/fs_wrapper.py:23]: why not reuse framework/io/fs.h?

File: python/paddle/fluid/tests/unittests/test_fleet_checkpoint.py [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+
+class FleetTest(unittest.TestCase):
+    def _test_check_point(self, fs, dir_path):
+        file_name = "persistables"
+
+        os.environ["TRAINING_ROLE"] = "TRAINER"
+        os.environ["PADDLE_TRAINER_ID"] = "0"
+        os.environ["PADDLE_TRAINER_ENDPOINTS"] = "127.0.0.1:6070"
+
+        #fs=LocalFS()
```
Comment (gavin1332) [inline python/paddle/fluid/tests/unittests/test_fleet_checkpoint.py]: remove commented code

File: python/paddle/fluid/tests/unittests/test_fleet_checkpoint.py [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+
+        exe = fluid.Executor(fluid.CPUPlace())
+        exe.run(fluid.default_startup_program())
+
+        status = TrainStatus(2)
+        fleet.save_check_point(exe, dir_path, train_status=status, fs=fs)
+        n1 = fleet._get_last_checkpoint_no(dir_path, fs=fs)
+
+        status2 = fleet.load_check_point(exe, dir_path, trainer_id=0, fs=fs)
+        assert status2 == status, "Checkpoint error!"
```
Comment (gavin1332) [inline python/paddle/fluid/tests/unittests/test_fleet_checkpoint.py]: why prefer assert to self.assertEqual？

File: python/paddle/fluid/tests/unittests/test_fleet_checkpoint.py [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+
+        fleet.save_check_point(exe, dir_path, train_status=status, fs=fs)
+        n2 = fleet._get_last_checkpoint_no(dir_path, fs=fs)
+        assert n2 == n1 + 1, "checkpoint increment error:{} {}".format(n1, n2)
```
Comment (gavin1332) [inline python/paddle/fluid/tests/unittests/test_fleet_checkpoint.py]: why prefer `assert` to `self.assertEqual`？

File: python/paddle/fluid/incubate/fleet/collective/__init__.py [hunk 9]
```diff
@@ -42,6 +46,21 @@ def __init__(self):
         pass
 
 
+class TrainStatus(object):
```
Comment (gavin1332) [inline python/paddle/fluid/incubate/fleet/collective/__init__.py:49]: Is it necessary construct a persistable training status object to recording context like training steps, the offset of filelist?

File: python/paddle/fluid/incubate/fleet/collective/__init__.py [hunk 10]
```diff
@@ -125,7 +150,182 @@ def save_persistables(self, executor, dirname, main_program=None):
...
-        io.save_persistables(executor, dirname, main_program, None)
+        io.save_persistables(executor, dirname, main_program, filename=filename)
+
+    def _save_train_status(self, path, train_status):
+        d = {}
+        d["epoch_no"] = train_status._epoch_no
+
+        file_name = "{}/fleet_train_status".format(path)
+        with open(file_name, 'w') as f:
+            json.dump(d, f)
```
Comment (gavin1332) [inline python/paddle/fluid/incubate/fleet/collective/__init__.py:161]: OK, that's what i mentioned before about persistable training status

File: python/paddle/fluid/incubate/fleet/collective/__init__.py [hunk 11]
```diff
@@ -125,7 +150,182 @@ def save_persistables(self, executor, dirname, main_program=None):
...
+
+        r = TrainStatus()
+        if not os.path.isfile(file_name):
+            return r
+
+        d = {}
+        with open(file_name, 'rb') as f:
+            d = json.load(f)
+
+        assert "epoch_no" in d, "Can't find epoch_no in dict from train_status file:{}".format(
```
Comment (gavin1332) [inline python/paddle/fluid/incubate/fleet/collective/__init__.py:174]: Default value could be assigned when exceptions occur. No necessary to stop training.

File: python/paddle/fluid/incubate/fleet/collective/__init__.py [hunk 12]
```diff
@@ -125,7 +150,182 @@ def save_persistables(self, executor, dirname, main_program=None):
...
+            d)
+        r._epoch_no = d["epoch_no"]
+        assert r._epoch_no >= 0, "Data in checkpoint file is not valid:{}".format(
```
Comment (gavin1332) [inline python/paddle/fluid/incubate/fleet/collective/__init__.py:177]: the same to what i mentioned above

File: python/paddle/distributed/utils.py [hunk 13]
```diff
@@ -0,0 +1,450 @@
...
+    def __eq__(self, n):
+        return self.hdfs_ugi == n.hdfs_ugi and \
+            self.hdfs_name == n.hdfs_name and \
+            self.hdfs_path == n.hdfs_path
+
+    def __ne__(self, n):
+        return not self == n
+
+
+class Cluster(object):
```
Comment (gavin1332) [inline python/paddle/distributed/utils.py:55]: These objects do not seem to be `utils`. Is it better to aggregate them to a new file which reserves EDL related components?

File: paddle/fluid/framework/io/shell.cc (hunk 1 above)
Comment (gongweibao) [reply paddle/fluid/framework/io/shell.cc]: Done.

File: paddle/fluid/framework/io/shell.cc (hunk 2 above)
Comment (gongweibao) [reply paddle/fluid/framework/io/shell.cc]: Done.

File: python/paddle/distributed/cloud_utils.py (hunk 3 above)
Comment (gongweibao) [reply python/paddle/distributed/cloud_utils.py:67]: This scenario is the job running under cloud cluster not local.
They must use a defined port so others can find him.

File: python/paddle/distributed/fs_wrapper.py (hunk 5 above)
Comment (gongweibao) [reply python/paddle/distributed/fs_wrapper.py:23]: 1. Need an abstract interface to hide the local and remote-fs difference.
2. Every action should return the action result and try when they meet error until timeout.
3. Many of the code under fs.h can be deleted.

File: python/paddle/fluid/tests/unittests/test_fleet_checkpoint.py (hunk 6 above)
Comment (gongweibao) [reply python/paddle/fluid/tests/unittests/test_fleet_checkpoint.py]: Done.

File: python/paddle/fluid/tests/unittests/test_fleet_checkpoint.py (hunk 8 above)
Comment (gongweibao) [reply python/paddle/fluid/tests/unittests/test_fleet_checkpoint.py]: Done.

File: python/paddle/fluid/tests/unittests/test_fleet_checkpoint.py (hunk 7 above)
Comment (gongweibao) [reply python/paddle/fluid/tests/unittests/test_fleet_checkpoint.py]: Done.

File: python/paddle/fluid/incubate/fleet/collective/__init__.py (hunk 9 above)
Comment (gongweibao) [reply python/paddle/fluid/incubate/fleet/collective/__init__.py:49]: They can be added when the action is implemented.And there may be more fields we think about now and it should support the user-defined fields I think.

File: python/paddle/fluid/incubate/fleet/collective/__init__.py (hunk 10 above)
Comment (gongweibao) [reply python/paddle/fluid/incubate/fleet/collective/__init__.py:161]: Same as above.

File: python/paddle/fluid/incubate/fleet/collective/__init__.py (hunk 11 above)
Comment (gongweibao) [reply python/paddle/fluid/incubate/fleet/collective/__init__.py:174]: If a file exists, this field must be in it.

File: python/paddle/fluid/incubate/fleet/collective/__init__.py (hunk 12 above)
Comment (gongweibao) [reply python/paddle/fluid/incubate/fleet/collective/__init__.py:177]: Same as above.

File: python/paddle/distributed/utils.py (hunk 13 above)
Comment (gongweibao) [reply python/paddle/distributed/utils.py:55]: The cloud cluster use them also.

File: python/paddle/distributed/cloud_utils.py (hunk 3 above)
Comment (gavin1332) [reply python/paddle/distributed/cloud_utils.py:67]: OK

File: python/paddle/distributed/fs_wrapper.py (hunk 5 above)
Comment (gavin1332) [reply python/paddle/distributed/fs_wrapper.py:23]: OK

File: python/paddle/fluid/incubate/fleet/collective/__init__.py (hunk 11 above)
Comment (gavin1332) [reply python/paddle/fluid/incubate/fleet/collective/__init__.py]: OK. It seems as a protocal

File: python/paddle/distributed/utils.py (hunk 13 above)
Comment (gavin1332) [reply python/paddle/distributed/utils.py:55]: In my opinion, `split_string`, `atoi`, etc. are utils, while `Cluster`, `Pod` are something like `Program`,  `Block`, and they are in framework domain.

**Review by gavin1332 — APPROVED** [review]
LGTM

**Review by JepsonWong — APPROVED** [review]
LGTM for requirements.txt.

**Review by raindrops2sea — APPROVED** [review]

**Review by guru4elephant — APPROVED** [review]
LGTM

**Pull request merged**